### PR TITLE
Add —non-interactive flag for automated processes

### PR DESCRIPTION
**Pull request by matt-adam**
**Summary**

If yarn is used within an automated process (in a CI system, for example), interactive prompts are not desirable.  This adds a commandline flag to remove them.

This resolves https://github.com/yarnpkg/yarn/issues/2625.

**Test plan**

I intentionally misconfigured a package.json with an invalid version of async:

    "async": "^111.5.0",

1) Run 'yarn'.  Output is an interactive menu letting me select a valid version of async to install.
2) Run 'yarn --non-interactive'.  Output is:

yarn install v0.22.0-0
[1/4] 🔍  Resolving packages...
error Couldn't find any versions for "async" that matches "^111.5.0"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

**Pending Issues**

* I wasn't sure how to build an effective test for this option.  If you think this warrants a test I'd appreciate some suggestions.
* I do not know how to update the documentation.

### CODE REVIEW

File: src/config.js [hunk 1]
```diff
@@ -36,13 +36,16 @@ export type ConfigOptions = {
...
   networkConcurrency?: number,
+  nonInteractive?: boolean,
 
   // Loosely compare semver for invalid cases like "0.01.0"
   looseSemver?: ?boolean,
 
   httpProxy?: ?string,
   httpsProxy?: ?string,
 
+  nonInteractive?: ?boolean,
```
Comment (bestander) [inline src/config.js]: why 2 times?

File: src/resolvers/registries/npm-resolver.js [hunk 2]
```diff
@@ -43,7 +43,7 @@ export default class NpmResolver extends RegistryResolver {
     const satisfied = await config.resolveConstraints(Object.keys(body.versions), range);
     if (satisfied) {
       return body.versions[satisfied];
-    } else if (request) {
+    } else if (request && !config.nonInteractive) {
```
Comment (bestander) [inline src/resolvers/registries/npm-resolver.js]: Is it the only interactive place we have?

**Comment by rkistner** [comment]
Can this also be enabled by default when using `--frozen-lockfile`? In that case it's not possible to persist the user's choice anyway.

**Comment by bestander** [comment]
I wonder if --non-interactive is too a too broad name for this change.
As it only touches npm resolution.

File: src/config.js (hunk 1 above)
Comment (matt-adam) [inline src/config.js]: Simple mistake, I will remove one of them.

**Comment by matt-adam** [comment]
I'm happy to change the name for this option in whatever way you'd like.  I figured yarn may add additional interactive prompts in the future, and having a single option name was more predictable.

The only other place I saw a prompt was in upgrade-interactive, but I didn't think an explicitly-interactive command needed an option to disable interactivity.  I'm happy to add it to that command as well though if you'd prefer.

@rkistner I don't know anything about --frozen-lockfile, to be honest, so I'm not a great person to ask about this.  I am happy to enable non-interactive when frozen-lockfile is specified, if you are sure this is the correct behaviour.

**Comment by bestander** [comment]
I think it is good to expand --non-interactive for other interactive commands in the future then.
Let's keep this implementation like this then.

**Comment by bestander** [comment]
Thanks, @matt-adam

**Comment by Ramblurr** [comment]
What about `yarn login`. I can't login programmatically:

```
error An unexpected error occurred: "Can't answer a question unless a user TTY".
```

**Pull request merged**